pd/src/testnet.rs

```diff
@@ -42,28 +45,101 @@ pub fn parse_validators(input: impl Read) -> Result<Vec<TestnetValidator>> {
-        // Tendermint peers are expected to maintain long-term persistent identities
-        // in the form of a public key. Each peer has an ID defined as
-        // peer.ID == peer.PubKey.Address(), where Address uses the scheme defined in
-        // crypto package.
-        // the peer addresses need to match this impl: https://github.com/tendermint/tendermint/blob/f2a8f5e054cf99ebe246818bb6d71f41f9a30faa/internal/p2p/address.go#L43
-        // The ID is for the node being connected to, *not* the connecting node's ID.
-        .map(|(id, ip)| format!("{}@{}", id, ip))
-        .collect::<Vec<String>>()
-        .join(",");
-    format!(
-        include_str!("../../testnets/tm_config_template.toml"),
-        node_name, peers_string,
-    )
+/// Use a hard-coded Tendermint config as a base template, substitute
+/// values via a typed interface, and rerender as TOML.
+pub fn generate_tm_config(
+    node_name: &str,
+    peers: Vec<TendermintAddress>,
+) -> anyhow::Result<String> {
+    tracing::debug!("List of TM peers: {:?}", peers);
+    let moniker: Moniker = Moniker::from_str(node_name)?;
+    let mut tm_config =
+        TendermintConfig::parse_toml(include_str!("../../testnets/tm_config_template.toml"))
+            .context("Failed to parse the TOML config template for Tendermint")?;
+    tm_config.moniker = moniker;
+    tm_config.p2p.seeds = peers;
+    Ok(toml::to_string(&tm_config)?)
+}
+
+/// Construct a [tendermint_config::net::Address] from a `node_id` and `node_address`.
+/// The `node_address` can be an IP address or a hostname. Supports custom ports, defaulting
+/// to 26656 if not specified.
+pub fn parse_tm_address(node_id: &Id, node_address: &str) -> anyhow::Result<TendermintAddress> {
+    let mut node = String::from(node_address);
+    // Default to 26656 for Tendermint port, if not specified.
+    if !node.contains(':') {
+        node.push_str(":26656");
+    }
+    Ok(format!("{}@{}", node_id, node).parse()?)
 }
+
+/// Query the Tendermint node's RPC endpoint and return a list of all known peers
+/// by their `external_address`es. Omits private/special addresses like `localhost`
+/// or `0.0.0.0`.
+pub async fn fetch_peers(
+    node_address: &TendermintAddress,
+) -> anyhow::Result<Vec<TendermintAddress>> {
+    let hostname = match node_address {
+        TendermintAddress::Tcp { host, .. } => host,
+        _ => {
+            return Err(anyhow::anyhow!(
+                "Only TCP addresses are supported for Tendermint nodes in Penumbra"
+            ))
+        }
+    };
+    let client = reqwest::Client::new();
+    let net_info_peers = client
+        .get(format!("http://{}:26657/net_info", hostname))
+        .send()
+        .await?
+        .json::<serde_json::Value>()
+        .await?
+        .get("result")
+        .and_then(|v| v.get("peers"))
+        .and_then(|v| v.as_array())
+        .cloned()
+        .unwrap_or_default();
+
+    let mut peers = Vec::new();
+    for raw_peer in net_info_peers {
```

Comment:
Open to feedback on this for-loop. Should this logic be separated out into a function? That'd allow us to `.collect` on the results more tersely. Maybe as a TryFrom trait that allows creating a `TendermintAddress` from a `serde_json::Value`?

Reply:
I think it's probably OK, if we were going to clean it up I think a potentially better route would be to use the `tendermint-rpc` crate, which will give us [parsed data](https://docs.rs/tendermint-rpc/latest/tendermint_rpc/endpoint/net_info/index.html).